#23992 Api support

manneohrstrom opened the pull request:
Added API methods which allows an external user to control the breakdown programmatically.

The following methods were added:

### analyze_scene()

```python
def analyze_scene(self):
        """
        Runs the scene analysis and returns a list of scene items.
        
        Each item is represented by a dictionary with a number of keys to 
        describe the item. This method uses the same logic that the main UI
        uses to determine the list of files.
        
        Only files whose path corresponds to a template in the toolkit templates
        file will be detected. Files do not need to exist as publishes in Shotgun
        however if they do, this method will return basic Shotgun publish metadata
        for them.
        
        The two keys node_name and node_type are used to return a DCC-centric 
        "address" or representation which makes it possible to identify the path
        within the DCC. In for example Maya and Nuke, this will return the 
        node name and type. The logic for this is implemented in the hooks and 
        will vary between DCCs.
        
        Here is an example of what the return data typically looks like:
        
        {'fields': {'Sequence': 'aaa',
                    'Shot': 'aaa_00010',
                    'Step': 'Comp',
                    'eye': '%V',
                    'height': 1556,
                    'name': 'test',
                    'output': 'output',
                    'version': 1,
                    'width': 2048},
         'node_name': 'Read2',
         'node_type': 'Read',
         'path': u'/mnt/projects/climp/sequences/aaa/aaa_00010/Comp/publish/elements/test/v001/2048x1556/aaa_00010_test_output_v001.%04d.dpx',
         'sg_data': {'code': 'aaa_00010_test_output_v001.%04d.dpx',
                     'entity': {'id': 1660, 'name': 'aaa_00010', 'type': 'Shot'},
                     'id': 1424,
                     'name': 'test',
                     'published_file_type': {'id': 3,
                                             'name': 'Rendered Image',
                                             'type': 'PublishedFileType'},
                     'task': {'id': 4714, 'name': 'Comp', 'type': 'Task'},
                     'type': 'PublishedFile',
                     'project': {'id': 234, 'name': 'Climp', 'type': 'Project'},
                     'version_number': 1},
         'template': <Sgtk TemplatePath nuke_shot_render_pub_mono_dpx>}
        
        This method will attempt to connect to shotgun, but the number of calls made are
        constant and independent of the scene complexity.
        
        You typically use it like this:
        
        >>> import sgtk
        >>> e = sgtk.platform.current_engine()
        >>> items = e.apps["tk-multi-breakdown"].analyze_scene()
        
        :returns: List of dictionaries, see above for example.
        """
```

### calculate_latest_version

```python
    def calculate_latest_version(self, template, fields):
        """
        Given a template and some fields, return the highest version number found on disk.
        The template key containing the version number is assumed to be named {version}.
        
        This will perform a scan on disk to determine the highest version.
        
        You typically use this in conjunction with the analyze_scene() method. 
        For example:
        
        import sgtk
        e = sgtk.platform.current_engine()
        breakdown_app = e.apps["tk-multi-breakdown"]
        
        # get list of breakdown items
        items = breakdown_app.analyze_scene()
        
        # now loop over all items
        for i in items:
        
            # get the latest version on disk
            latest_version = breakdown_app.calculate_latest_version(i["template"], i["fields"])
            
            # if our current version is out of date, update it!
            current_version = i["fields"]["version"]
            if latest_version > current_version:
                
                # make a fields dictionary representing the latest version
                latest_fields = copy.copy(i["fields"])
                latest_fields["version"] = latest_version
                
                # request that the breakdown updates to the latest version
                breakdown_app.update_item(i["node_type"], i["node_name"], i["template"], latest_fields)
         
        :param template: Template object to calculate for
        :param curr_fields: A complete set of fields for the template
        :returns: The highest version number found
        """
```

### update_item

```python
    def update_item(self, node_type, node_name, template, fields):
        """
        Request that the breakdown updates an given node with a new version.
        This is similar to running the update in the breakdown UI. The actual 
        update call will be dispatched to a hook which handles the DCC specific logic.
        
        You typically use this in conjunction with the analyze_scene() method. 
        For example:
        
        import sgtk
        e = sgtk.platform.current_engine()
        breakdown_app = e.apps["tk-multi-breakdown"]
        
        # get list of breakdown items
        items = breakdown_app.analyze_scene()
        
        # now loop over all items
        for i in items:
        
            # get the latest version on disk
            latest_version = breakdown_app.calculate_latest_version(i["template"], i["fields"])
            
            # if our current version is out of date, update it!
            current_version = i["fields"]["version"]
            if latest_version > current_version:
                
                # make a fields dictionary representing the latest version
                latest_fields = copy.copy(i["fields"])
                latest_fields["version"] = latest_version
                
                # request that the breakdown updates to the latest version
                breakdown_app.update_item(i["node_type"], i["node_name"], i["template"], latest_fields)
        
        :param node_type: Note type of the item to update, as returned by analyze_scene()
        :param node_name: Note name of the item to update, as returned by analyze_scene()
        :param template: Template object representing the path pattern to update
        :param fields: Fields dictionary representing the values to apply to the template in order
                       to render an valid and existing path on disk that the system can update to.
         """
```

alandann commented on app.py:
```diff
@@ -44,4 +44,176 @@ def show_breakdown_dialog(self):
+        # if shotgun data is returned for an item, trim this down
+        # to return a more basic listing than the one returned
+        # from get_breakdown_items:
+        for i in items:
```
we should avoid single character variable names

alandann commented on app.py:
```diff
@@ -44,4 +44,176 @@ def show_breakdown_dialog(self):
+        For example:
+        
+        import sgtk
+        e = sgtk.platform.current_engine()
```
is 'e' used anywhere else (it's pretty hard to search for as it's a single character!) - if it's not then I'd be inclined to collapse these two lines into one and get rid of 'e', otherwise use more letters (e.g. 'eng') :)

alandann commented on app.py:
```diff
@@ -44,4 +44,176 @@ def show_breakdown_dialog(self):
+                breakdown_app.update_item(i["node_type"], i["node_name"], i["template"], latest_fields)
+         
+        :param template: Template object to calculate for
+        :param curr_fields: A complete set of fields for the template
```
param doesn't match param 'fields'

alandann commented on app.py:
```diff
@@ -44,4 +44,176 @@ def show_breakdown_dialog(self):
+        You typically use this in conjunction with the analyze_scene() method. 
+        For example:
+        
+        import sgtk
```
It seems really verbose having the same full example in the doc-string of both of these methods?  Maybe instead the example could move to the doc-string for the app/class and the examples in these methods could be the specific piece that uses the method?

alandann commented on app.py:
```diff
@@ -44,4 +44,176 @@ def show_breakdown_dialog(self):
+                # request that the breakdown updates to the latest version
+                breakdown_app.update_item(i["node_type"], i["node_name"], i["template"], latest_fields)
+        
+        :param node_type: Note type of the item to update, as returned by analyze_scene()
```
Is 'Note' supposed to be 'Node'?

alandann commented on app.py:
```diff
@@ -44,4 +44,176 @@ def show_breakdown_dialog(self):
+                breakdown_app.update_item(i["node_type"], i["node_name"], i["template"], latest_fields)
+        
+        :param node_type: Note type of the item to update, as returned by analyze_scene()
+        :param node_name: Note name of the item to update, as returned by analyze_scene()
```
and here

alandann commented on python/tk_multi_breakdown/breakdown.py:
```diff
@@ -0,0 +1,236 @@
+                # store the normalized fields in dict
+                items.append(item)
+
+    # now now do a second pass on all the files that are valid to see if they are published
```
now now

alandann commented on python/tk_multi_breakdown/breakdown.py:
```diff
@@ -0,0 +1,236 @@
+        # cache item
+        g_cached_sg_publish_data[path] = sg_chunk
+
+        # change type from valid -> publish
```
this comment doesn't match the code!

alandann commented on app.py:
```diff
@@ -44,4 +44,176 @@ def show_breakdown_dialog(self):
+        
+        Here is an example of what the return data typically looks like:
+        
+        {'fields': {'Sequence': 'aaa',
```
I wonder if the app API needs to return all this data - it really feels like it's exposing a lot of the internal functionality to the caller and also makes them understand this data before they can use the other methods.  Would this be too simple:

```python
# get a list of items (each item in the list is dictionary containing {'node type', 'node name', 'path', 'version'}
items = app.analyse_scene()

for item in items:
    # calculate the latest version:
    latest_version = app.calculate_latest_version(item["path"])

    # if latest is newer then update:
    if item["version"] < latest_version:
        app.update_item(item["node_type"], item["node_name"], item["path"], latest_version)
```
?

alandann commented:
Done - last comment is probably the main thing - I wonder if the data returned from the `analyse_scene()` call and how that makes the arguments required for the other methods is too complex?

manneohrstrom commented:
Addressed all the comments.

As for the last question, yes I did consider back and forth and I think all the data is needed. I dropped the path, which is redundant and can be retrieved via the template system (template+fields).

- Node name/node type seems like a sensible concept for generically identify objects in a DCC DAG (or equivalent) so that's needed
- template/fields is needed to identify the file or resource. By dropping that path (which i just did), we actually make it path agnostic - you could imaging adding a new type of templates (say something cloud pased), where that template object + fields resolve into a different structure. By not returning the path and just returning template object + fields, we are allowing for new future template types to be transparently supported.
- the shotgun data was the thing I personally wasn't sure about. Here is my reasoning: It seems that the whole point of this breakdown is to gracefully connect into the world of shotgun and provide shotgun metadata about the objects. We want the "what is latest" be driven by shotgun data in the future, not just scan for the latest thing on disk. With all this in mind, I always saw the shotgun data as a key asset to return - and the case where we have an unpublished asset being the edge case. In terms of the shotgun fields, I narrowed it down to the "system" fields, so those should all be non-controversial to return. The rationale here is that by returning all the common fields, the client user can avoid making it's own separate query to shotgun to get stuff (which would have been the case if we just returned type/id).

So in summary, I see your concern but I think it's ok - it looks substantial but it's actually just a bunch of pretty basic fields!

We also need to consider future upcoming changes somewhat – these include swapping of references, ability to update not just to the latest version (but to latest approved, with logic using shotgun to determine this) and ability to lock certain assets (to name a few). The current approach with the scan method, the calculate latest and update methods is trying to cater for these use cases. By passing fields/template to the update method rather than a path and a version number for example, we make it easier for clients to build a LOD swapping system based on this (a common request). Generally speaking, I found that passing a template+fields pair generally is a good way to pass the notion of a file resource around!

Hope it makes sense and thanks for the awesome feedback!

beauroberts commented:
Long-time lurker, first-time commenter!

Just thought I'd flag that "analyse" is the British English spelling, and analyze is the North American English spelling, in case you have standardised (or standardized?) around the latter!

manneohrstrom commented:
Oooooh!! Mr Roberts! Awesome! Warmly welcome!

Yes UK vs US english is a constant source of confusion for me! In the code, the idea is to use US english throughout. I too was greatly confused when alan started to say "analyse" in the comments - i went in and double checked that the code was using US and not UK spelling.

The fact that "analyze" gets underlined in red by a UK computer as i type doesn't exactly help :)

jfboismenu-adsk commented on app.py:
```diff
@@ -44,4 +44,153 @@ def show_breakdown_dialog(self):
+        
+        The two keys node_name and node_type are used to return a DCC-centric 
+        "address" or representation which makes it possible to identify the path
+        within the DCC. In for example Maya and Nuke, this will return the 
```
For example in Maya and Nuke

jfboismenu-adsk commented on app.py:
```diff
@@ -44,4 +44,153 @@ def show_breakdown_dialog(self):
+        :returns: The highest version number found
+        """
+        tk_multi_breakdown = self.import_module("tk_multi_breakdown")
+        return tk_multi_breakdown.compute_highest_version(template, fields)
```
Wouldn't a failed publish (the file is created but not registered with Shotgun) be returned if it was the highest version, resulting in a false positive?

jfboismenu-adsk commented on python/tk_multi_breakdown/breakdown.py:
```diff
@@ -0,0 +1,240 @@
+    return items
+
+
+def compute_highest_version(template, curr_fields):
```
Given a scene with hundreds or thousands of reference to the same file (a crowd for example or a tree in a forest), you will be scanning files on disk forever, especially if there are multiple versions for that file. Seems like some caching is in order.
